### ISSUE TITLE
Make the underlying color_list type for matrix layouts useful by slicing out the driver-oriented part.

### ISSUE THREAD

**Issue by rec** [ManiacalLabs/BiblioPixel#456]
This overlaps a couple of other issues, but I wanted to get my idea out about this.

These days, I see 95% of our programmers as wanting to write `Animations`.  

I see very few new features that would result in new `Layouts`, and new `Drivers` are always extremely desirable, but are going to involve advanced programmers!  Most people who simply have a quick project in mind will be making a new `Animation`.

So I want to make life easy and fast for them, even if it requires more work from us.

---

Since the start, a big part of what makes BP useful is that it compensates automatically for wiring issues - rotate, serpentine and reflections, the "wiring parameters".

This is a prime, important feature - one we have to keep, and one that's "complete" (i.e. we cover all reasonable wiring cases...)  

Unfortunately, it's implemented in the wrong place!  :-)

It's implemented as a get/set wrapper in `layout.matrix.Matrix` which takes incoming "geometric" coordinates and transforms them into "device coordinates" - using the wiring parameters to do so.

What it means is that all the incredibly power of the underlying `color_list` - particularly slices, or conceivably fast numpy features if they are using numpy - is rendered useless. 

For example, look at my [best try](https://github.com/rec/BiblioPixel/blob/a631dd3589fd04099e9b77c61476dfe40d2197d7/bibliopixel/animation/reprocess.py#L23-L57) at the code for the sorting animation.

It's really slick.  The program extracts a slice for each row or column in the matrix, then uses that slice to extract entries into a list.  Then client code gets to make a pass at that list - i.e. "do one sorting step" - and then the same slice is used to store those elements back where they came from!

More, this could be made significantly faster by caching the slices (which never change) and the lists of elements (which always have the same length and number).

And it works great!  _If_ (you saw this coming) the underlying `layout.Matrix` is not serpentine.

(Don't worry, fix coming.  You'll love the results!  But I'm going to keep the old slice code around against a better day - there are tests for it...)

---

So how _should_ this be working?  

Conceptually, the same thing has to happen - API user writes to a point (x, y) (or (x, y, z), etc) and eventually a pixel at some different point (x', y') is actually written to, depending on the wiring parameters, serpentine and all her buddies. 

We all agree here.

However, doing this computation on each access to the underlying color_list is, I argue above, not the place to do this.

Instead, we need to do this during the _next_ phase of computation - the one where we take the color_list and generate a `bytearray` from it that we send to the hardware!

I propose that _in code_ we extract out all the wiring parameters as a separate class called `Transform`.  Projects and existing code would not change for now - this would be a class owned by the Layout.

The driver wouldn't have any more work to do.  It would get exactly the same data as it got now, but with one extra optional parameter - the transform, which it can use to extract data from the color_list.

This is the perfect place for this.  We're doing pretty heavy computations on each pixel here anyway - why not simply start by getting a _different_ pixel from a fixed list?

**Backwards compatibility**

It would be easy to maintain perfect backwards compatibility here, as long as existing code wasn't directly manipulating the `color_list` (but you're not supposed to, and if you are, you're probably screwing up the wiring parameters like I did in that code sample...)

For the moment, we'd keep the old layout features, but also have an optional Transform.  If Transform were specified, we'd turn off the wiring parameters and use the transform.

Old-style drivers would work exactly the same.  To take full advantage of `Transform`'s efficiency boost, you'd have to rewrite your code a bit, but it would run at the old speed just fine without changes.

**Stretch goal for the future**

Once we did this, we could expose the color_list and let people use it directly.

Today our color lists are exactly (i, 3) dimensional numpy arrays and lists of tuples

For Matrices and Cubes, why couldn't we directly support numpy (x, y, 3) and (x, y, z, 3) arrays - in parallel of course with today's color_list?

The code would be short and easy, once the above is done... it'd simply be another transformer!

Imagine how easy it would be for our users to say:

    frame = 0.3 * f1 + 0.7 * f2

when `f1` and `f2` were actually big 3- or 4-dimensional arrays?

It means the full wealth of numpy and scipy tools are directly available to them for free.

**Comment by adammhaile** [comment]
Sounds good to me :)
